### DISCUSSION TITLE
Update: GUI Progress...

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#451] [General]
I wanted to share some details on the GUI progress being made.  Please remember that we're initially working on custom drawn controls at first, but a lot of this is designed to work with the Windows based controls too (which will be 100% VBx compatible).

First up, let's look at  some visuals:

<img width="1075" alt="tbCustomControlsProgress1" src="https://user-images.githubusercontent.com/25738109/134643747-cce5bf49-9a85-4a32-b6ce-2c80f5ca534c.png">

Don't worry that the property sheet looks complicated here... this is just to demonstrate the flexibility of the property sheet, which is powered from the public properties of the WaynesButton class definition.  The final built-in button class will be a little simpler in design!

All of the buttons pictured above are drawn from the following CustomControl defined in twinBASIC:

<img width="491" alt="tbCustomControlsButtonClass" src="https://user-images.githubusercontent.com/25738109/134643949-9276baf2-070e-4a4c-86e6-21abe7ed1323.png">
<img width="459" alt="tbCustomControlsButtonState" src="https://user-images.githubusercontent.com/25738109/134643956-a0164737-ddd0-4216-9ba4-f763eb543b8f.png">

As you can see, designing a custom control in twinBASIC will be very easy.   The main method of interest is the OnPaint implementation that in this instance just passes on the descriptions of the Corners, Borders, Background etc to the DrawElement method, and then the twinBASIC form engine handles the rest.

For those that are interested in how the forms and control properties are persisted; it is JSON as pictured here...

<img width="470" alt="tbCustomControlsFormSerialization" src="https://user-images.githubusercontent.com/25738109/134644954-7c30fd47-3b96-414e-ba0e-d192657dfbf2.png">

**Comment by bclothier** [comment]
I love the fact that property sheet allows you to add new properties. I hope it will also let us rename the property name --- it's nice to have `BottomRightCornerColor` or `DarkColor` than `Values(4)`. 

I wonder if that implies that tB will need to improve how it handles arrays? In VBx, I find myself avoiding arrays like plague because they are unwieldy to work with when you have unknown numbers of members and `Redim Preserve` is just a waste of cycle. 

Coming along nicely!

**Reply by WaynePhillipsEA** [reply]
Thanks Ben.  Indeed, the corners are not arrays, but named as you might expect.   Apologies as I couldn't fit everything on the one screenshot.
<img width="382" alt="tbCorners1" src="https://user-images.githubusercontent.com/25738109/134687895-4b5a48b4-2565-4054-87f4-9a39cc7c3b3d.png">

If you look closely, the Values(0...4) array in the first screenshot are the color points of the gradient for the BackgroundFill.  You can define as many color points as you like to make complex gradients, and so an array is the most sensible way of storing them.   You don't necessarily have to expose that level of detail in the property sheet though(!)... you could instead have the BackgroundFill as a *private* field and then it won't appear in the property sheet, and then you could just populate the color points array via your own property getters/setters as you see fit.  Your property getters/setters will be exposed on the property sheet if they're Public.    Also note however that there is a nice gradient picker included for setting the gradient color points without messing with the array Values() anyway (as pictured).

<img width="555" alt="tbColorPoints1" src="https://user-images.githubusercontent.com/25738109/134689923-6869073a-981e-4596-8930-61e55630013b.png">

<img width="322" alt="tbColorPoints2" src="https://user-images.githubusercontent.com/25738109/134690151-9822915d-fed1-4d36-8fde-718797a0b573.png">
(each color point also allows for alpha channel, with the slider as shown above)

**Comment by bclothier** [comment]
Few questions... 

Do I perceive correctly that a property exposing an object will in turn expose the object’s properties as a subtree in the property sheet? That would be a huge improvement over the flat sheet VBx uses.

Do you have a plan for providing an uniform and extensible system of serialization? I see that `ControlContext` implements `Deserialize`, but does the `WayneButton` get to control its serizalization? In VBx world, it’s not possible to implement this without writing a _lot_ of code. The .NET system does a pretty good job of making it easy to create and customize serizaliable classes. Would love to see the same with tB.

I’m wondering why is `ButtonState` newing up a `Canvas`? I would have expected it to get it from `WayneButton`?

**Reply by WaynePhillipsEA** [reply]
Yes, that's right.  The property sheet works natively with all the basic types including objects, UDTs, enums and arrays (including arrays of Objects/UDTs as well).  The only exception is that it doesn't support Variants, due to the need to represent both a type and a value in the property sheet, which is just awkward and unnecessary IMO... just avoid Variants in public properties of custom controls.

Something I didn't mention before;  you can also extend the property sheet with custom editors.  You can control the custom HTML div for any particular exposed property (specified with a tB attribute and associated javascript to edit the DOM).  You can see this in the screenshots for 'Color' properties and for the ellipses '...' button for ColorPoints that launches the gradient picker popup (Grapick).

>Do you have a plan for providing an uniform and extensible system of serialization?

The deserialization needs some more thought to allow for customization.  At the moment, it's a black box, with the Deserialize call handling all the nitty gritty details, but not offering any customization.  I'm certainly open to ideas on this.

>I’m wondering why is ButtonState newing up a Canvas? I would have expected it to get it from WayneButton

This might change before you get your hands on it.  This was originally designed so that each button state (normal, pressed, focused etc) could pre-render itself to its own buffered canvas, however that caching could cause issues (e.g. on multiple DPI monitor situations), and also adds CPU and memory pressure for what would be only a minor performance benefit when switching between button states at runtime.

**Comment by WaynePhillipsEA** [comment]
The class design has now changed somewhat:
<img width="426" alt="tbButtonv2" src="https://user-images.githubusercontent.com/25738109/134795704-8e8f5f2a-fdad-4302-b88b-aa4f4a78c861.png">

CustomControls now need to implement ICustomControl, and this provides for Initialize/Destroy/Paint/MouseDown events all in one place (more like traditional ActiveX control design in VB6).

In light of the Initialize event, the serialization constructor has been decoupled from the CustomControlContext.  In future, when we support more refined serialization (see #453), it will now be easy to switch to use the alternative serialization constructor (TBD) if you need more refined control over the serialization process.

**Comment by WindowStations** [comment]
Looks great.  Are there any standard accessibility options implemented or planned?  For example screen reader support.  I just wanted to touch base on that because there are work environments that require this functionality by law.  Another possibility is accepting XInput which automatically operates with the xbox adaptive controller as well as the standard controller.

**Reply by WaynePhillipsEA** [reply]
Certainly planned, but not implemented as yet.  The first iterations will be to get the general development framework in place so that devs can experiment with the capabilities and I can get some feedback.  Then I'll start focusing a little more on these finer details.

**Reply by WindowStations** [reply]
Understood, it's great to see this unfolding so well...  I still have not got a chance to play with the extensibility potential, but I know that "side by side" interop should be possible with tB.  I've been making significant progress with the undocumented VBA6.dll API.  I know that's not in the long term cross platform vision, but for 100 (virtual) percent compatibility it will be invaluable.  Thanks for all of the hard work!

**Comment by dmrvb** [comment]
So self drawn controls written in TB? That's excellent.
This should make 64bit and cross platform a lot easier.
You will likely need to support the existing OCX controls to give full compatibility, but if these new controls can have full functionality (with drag drop etc etc) then they should be the preferred controls in the future?.
Once you have done the foundations could the community here perhaps contribute some of the "big" controls like TreeView, ListView and RichTextBox????
You are providing multiple visual styles but again if you need help could others do some of this????
Looks like we will have extensive flexibility with control appearance but if we really want will we be able to customise the actual drawing code ourselves?.

**Reply by WaynePhillipsEA** [reply]
Yes, exactly.  These are designed specifically with cross platform in mind, but I suspect the very clever devs on here will be able to replicate the feel of typical Windows controls very effectively.   I'm putting together some samples over the next few days (e.g. a simple datagrid), and then I'm going to pass things over to the community here to see what happens whilst I switch to working on 64-bit support for a couple of weeks.

>Looks like we will have extensive flexibility with control appearance but if we really want will we be able to customise the actual drawing code ourselves?.

The main drawing method is DrawElement, which is extremely flexible and is the only method used for painting the buttons and text elements in the screenshots provided earlier in this post.  After that there is also DrawLine and SetPixel methods for custom drawing.  Other drawing methods will be added in due course, but a lot of controls can already be created with just this limited set of drawing methods.

**Comment by WaynePhillipsEA** [comment]
Here is a video of the form designer in action, and a simple grid control (less than 100 lines of code for the OnPaint method).   The first beta of this feature will be available to those that have pre-ordered twinBASIC in the coming week.

[![twinBASIC: CustomControls and form designer sneek peek!](https://res.cloudinary.com/marcomontalbano/image/upload/v1633275644/video_to_markdown/images/youtube--nC7r4JdwB_A-c05b58ac6eb4c4700831b2b3070cd403.jpg)](https://youtu.be/nC7r4JdwB_A "twinBASIC: CustomControls and form designer sneek peek!")

**Reply by bclothier** [reply]
That's positively great stuff you've got going there, Wayne!

Few follow up questions if I may....

1) I see that we can define how to paint on the canvas but how do we indicate that it should react to a mouse click, mouse hovering, or to some kind of keyboard inputs? That would be a prerequisite for us to be able to create events for the custom controls, right? 

2) Will there be support for expressions in property sheet? In Access and PowerApps, you can do something like `=CallAComplicatedFormula()` which has its ups and downs. My biggest problem with using it in Access is that there's no way to track the references, but it's also much simpler and tidier to be able to use expressions for a property than having to run an unrelated piece of code to mutate the property.

3) With the traditional ActiveX (and I think also the WinForms) controls, it was necessary to make a distinction between runtime and design-time. How will this work out with the twinBasic custom controls? Do we need to be careful to not run runtime code during the design-time and vice versa?

**Reply by WaynePhillipsEA** [reply]
Thanks @bclothier!

1... working on that this week.  There will be raw events passed through the ICustomControl for MouseDown, MouseMove, KeyDown etc.  However, these could be difficult to work with when you think about things like the curved corners in the painting, since how would you, as the developer of a CustomControl know whether the mouse is actually in a painted area, and not in a transparent part of the image (like a transparent corner section)?   So my plan is to allow DrawElement to also register callbacks for things like MouseMove/MouseDown, so that the actual painted element is directly associated with a callback event.  I believe this will be much simpler than handling mouse events manually. 

2 I did consider this, but at the moment it's not planned.  In order to implement it like Access does, we'd have to link in big parts of the compiler at runtime in order to offer a proper runtime expression evaluation service.  A compromise might be to only allow the expressions to be set at compile-time, and so the expressions are effectively compiled-in to the form design.  That would also solve the tracking of references problem, since the expression then forms part of the codebase and would be considered as source code for refactoring operations.

3 Yes, I'm sure it will be necessary in some instances, and so a property will be provided for determining this.

**Reply by bclothier** [reply]
Wayne,

Regarding 2...

💡 I like what you're suggesting. You could do something like:

```
Public MySuperDuperProperty As SomeFancyPantsWidget = DoSomeComplicatedCalculationsYo()
```

Thus, one can put in an expression in the property sheet which would then reference a defined procedure, which would be equivalent of mutating the defining property in the source code. The downside is that one can't do something like this:

```
Public Sub Command1_Click()
'Let's change the expression for giggles
Me.MySuperDuperControl.MySuperDuperProperty = "=DoAnotherThing()"
End Sub
```

But I think that is easily worked around by having the expression take a parameter or handling it within its body.

Regarding 3....

While we might need the property, I never particularly liked the idea of having a single click event handler run in both design-time and runtime with branching. It does not feel clean to me. Were it up to me, I'd probably provide a separate `ICustomControlDesigner` interface which can be then optionally implemented to provide design-time experience and thus keep the design-time and runtime code neatly separated.

**Reply by mwolfe02** [reply]
From @WaynePhillipsEA's comment (emphasis and link mine):
> 2 I did consider this, but at the moment it's not planned. In order to implement it like Access does, we'd have to link in big parts of the compiler at runtime in order to offer a proper runtime expression evaluation service ([Expressions vs. Code](https://nolongerset.com/expressions-vs-code/)). **A compromise might be to only allow the expressions to be set at compile-time, and so the expressions are effectively compiled-in to the form design**. That would also solve the tracking of references problem, since the expression then forms part of the codebase and would be considered as source code for refactoring operations.

I really like the idea of supporting compile-time expressions in control properties.  The reference-tracking problem is a real issue in Microsoft Acccess, and one that I did not fully appreciate until I had expressions all over the place in my codebase (see [Lightweight Forms? Just Don't](https://nolongerset.com/lightweight-forms-just-dont/)).

**Comment by dmrvb** [comment]
Drawing of these controls is interesting, some new graphics methods would be great but need some thought..
GDI is fast and powerful, but a bit perverse, and not cross platform.
The VB6 drawing commands can be slow and get a lot of criticism, I assume you will have to replicate them for compatibility, but a new drawing system would be really good for the future, so these control drawing commands could become the basis of the new cross platform graphics system.

**Comment by mwolfe02** [comment]
How is the form design JSON generated?  

I'm asking because of its impact on version control.  For example, are the controls arranged:

- **alphabetically** by name
- "**geographically**" by location on the form
-  **sequentially** in the order they are added to the form (I believe this is how Access forms are generated when `SaveAsText`ed)

Ideally, they would be generated in such a way as to [minimize the noise](https://nolongerset.com/signal-vs-noise/) within version control.  For example, changing a single property should not affect more than a single line in the JSON.  

With that in mind, I think the best approach would be to arrange the controls **sequentially**, but I think it would be a discussion worth having because there may be other considerations I'm not thinking of.

**Reply by mwolfe02** [reply]
The other thing that Access does in its exported form syntax is to store the control defaults for that form in a header section.  Then, for each individual control, only properties with a non-default value are explicitly saved.  For example, boolean properties are not exported with a Yes/No, True/False, or 0/1.  Instead, they are exported with a value of `NotDefault` or they are not exported at all.  

I don't know that this approach is good or bad--there are always tradeoffs--I just thought I'd point it out as something to consider (if you haven't already).

**Reply by WaynePhillipsEA** [reply]
Thanks @mwolfe02.   Controls themselves are stored in arrays based on the creation order (sequentially as you describe), as this is important for maintaining z-ordering and docking-ordering.  The properties of controls are stored in alphabetical order in order to minimize noise in case of changes or updates to control property sets.
